### PR TITLE
Voltage anti-islanding detection

### PR DESCRIPTION
**Pull request by exfletch**
Changes:

- Updated database build process to handle optional columns without changing script every time
  - optional columns are max/min/mean voltage/frequency
- Added detection of anti-islanding threshold excursions
  - Flag to indicate excursions have occurred is included in circuit summary and plot aggregation options
  - Per-excursion summary is generated, save button added to UI

### CODE REVIEW

**Review by phoebeheywood — COMMENTED** [review]
All looks pretty good to me, just left a few small suggestions. Two other things
- Sites with voltage always 0 are classified as undervoltage, so may be worth adding in a check for that and replacing with NA
- The voltage excursion summary doesn't filter for tesla islanded sites so shows more sites than are in the circ sum when the "exclude islanded sites" toggle is on

File: run_analysis/run_analysis.R [hunk 1]
```diff
@@ -354,15 +362,15 @@ run_analysis <- function(data, settings) {
...
+        combined_data_cols <- c("ts", "site_id", "c_id", "power_kW", "c_id_norm_power", "v", "vmin", "vmax", "vmean", "f", "s_state",
                                 "s_postcode", "pv_installation_year_month", "Standard_Version", "Grouping", "sum_ac",
                                 "clean", "manufacturer", "model", "site_performance_factor", "response_category",
                                 "zone", "distance", "lat", "lon", "e", "con_type", "first_ac", "polarity",
                                 "compliance_status", "reconnection_compliance_status",
                                 "manual_droop_compliance", "manual_reconnect_compliance", "reconnection_time",
                                 "ramp_above_threshold", "c_id_daily_norm_power", "max_power", "ufls_status",
                                 "pre_event_sampled_seconds", "post_event_sampled_seconds", "ufls_status_v",
-                                "pre_event_v_mean", "post_event_v_mean")
+                                "pre_event_v_mean", "post_event_v_mean", "vmin_na", "vmax_na", "vmean_na", "antiislanding_v_excursion_2015")
```
[the remark this answers is not in the document]
Comment (exfletch) [reply run_analysis/run_analysis.R:373]: Nope good find thanks!

File: BDInterface/interface.R [hunk 2]
```diff
@@ -18,8 +18,8 @@ DBInterface <- R6::R6Class("DBInterface",
   public = list(
     db_path_name = NULL,
     default_timeseries_column_aliases = list(ts='_ts', time_stamp='_ts', c_id='_c_id', v='_voltage', vmin='_vmin', 
-                                             vmax='_vmax', f='_frequency', fmin='_fmin', fmax='_fmax', e='_e', d='_d', 
-                                             p='_p'),
+                                             vmax='_vmax', vmean='_vmean', f='_frequency', fmin='_fmin', fmax='_fmax',
```
[the remark this answers is not in the document]
Comment (exfletch) [reply BDInterface/interface.R:21]: Probably a bit too large a scope to include here, but worth adding to the list of tasks I think.

File: BDInterface/interface.R [hunk 3]
```diff
@@ -246,21 +248,25 @@ DBInterface <- R6::R6Class("DBInterface",
...
-        from file_con"
-      } else{
+      optional_columns <- c("vmin", "vmax", "vmean", "fmin", "fmax")
+      replace_columns <- paste(optional_columns[optional_columns %in% column_names], collapse=", ")
+      select_columns <- c()
+      for (col in optional_columns[optional_columns %in% column_names]) {
+        select_columns <- c(select_columns, sprintf("_%s as %s", col, col))
+      }
+      select_columns <- paste(select_columns, collapse=",\n")
+      if (any(optional_columns %in% column_names)){
```
[the remark this answers is not in the document]
Comment (exfletch) [reply BDInterface/interface.R:253]: Good suggestion. We still need to check if the optional columns exist, but can use that to determine whether to include the comma or not.

File: anti_islanding_detection/anti_islanding.R [hunk 4]
```diff
@@ -0,0 +1,117 @@
...
+        (vmax > OVERVOLTAGE_1_2020) & in_event_window, "overvoltage_1", antiislanding_v_excursion_2020))
+    combined_data <- mutate(combined_data, antiislanding_v_excursion_2020=ifelse(
+        (vmax > OVERVOLTAGE_2_2020) & in_event_window, "overvoltage_2", antiislanding_v_excursion_2020))
+
+    return(combined_data)
+}
+
+#' Find the number of consecutive occurrences of each antiislanding threshold crossing
+#' @param antiislanding_column: the column containing which threshold has been crossed
+#' @param recurrances_columm: the name of the column for the output count
```
[the remark this answers is not in the document]
Comment (exfletch) [reply anti_islanding_detection/anti_islanding.R]: fixed

File: BDInterface/interface.R [hunk 5]
```diff
@@ -159,6 +160,7 @@ DBInterface <- R6::R6Class("DBInterface",
       RSQLite::dbExecute(con, "UPDATE timeseries SET v = NULL WHERE v = ''")
```
[the remark this answers is not in the document]
Comment (exfletch) [reply BDInterface/interface.R]: Done

**Comment by exfletch** [comment]
> * Sites with voltage always 0 are classified as undervoltage, so may be worth adding in a check for that and replacing with NA

Do you think that replacing this with NA/NaN is OK? Are there any datasets that can operate in island mode but report zero grid voltage validly?
What threshold would you reckon we should use for this?

> * The voltage excursion summary doesn't filter for tesla islanded sites so shows more sites than are in the circ sum when the "exclude islanded sites" toggle is on

Have done this

File: BDInterface/data_cleaning_functions.R [hunk 6]
```diff
@@ -185,4 +185,33 @@ check_for_mixed_polarity <- function(combined_data){
   combined_data <- mutate(combined_data, con_type=ifelse(max_power > ac * .1 & min_power * -1 > ac * 0.1, 
                                                          "mixed", con_type))
   return(combined_data)
-}
\ No newline at end of file
+}
```
Comment (exfletch) [inline BDInterface/data_cleaning_functions.R:188]: @phoebeheywood the following addresses low voltage concerns by setting permanent outliers to NaN

**Review by phoebeheywood — APPROVED** [review]
All looking and working well for me now, thanks Jack :)

**Pull request merged**